torch/csrc/utils/tensor_numpy.cpp

```diff
@@ -386,6 +396,314 @@ at::Tensor tensor_from_cuda_array_interface(PyObject* obj) {
+  /*
+  A utility function for returning a string representation of a std::vector containing ints/floats.
+  Useful for printing the shape of an N-dimensional tensor. This is used in the error message creation
+  part in the _extract_ndarrays function below.
+  */
+  std::ostringstream output;
+  output << "(";
+
+  if (end < 0) {
+    end = vec.size();
+  }
+  for (int64_t i = start; i < end; ++i) {
+    if (i != end - 1) {
+      output << vec[i] << ", ";        
+    } else {
+      output << vec[i] << ")";
+    }
+  }
+  return output.str();
+}  
+
+template <typename T>
+std::string get_buffer_str(void* buffer, int64_t start, int64_t end) {
+  /*
+  A similar utility function as get_vector_str, except it works on an array buffer instead of vector.
+  */
+  T* array = static_cast<T*>(buffer); 
+  std::ostringstream output;
+  output << "(";
+
+  for (int64_t i = start; i < end; ++i) {
+    if (i != end - 1) {
+      output << array[i] << ", ";        
+    } else {
+      output << array[i] << ")";
+    }
+  }
+  return output.str();
+}  
+
+inline bool _shape_match_ndarray(std::vector<int64_t> & shape, PyArrayObject* array, int64_t start) {
+  /*
+  A utility function that returns true only if shape[start:] == array.shape where array is a numpy array.
+  Else returns false. This is used in _extract_ndarrays() below for checking whether an embedded numpy array
+  has the target shape.
+  */
+  int64_t ndim = PyArray_NDIM(array);
+  if ((shape.size() - start) != ndim) {
+    return false;
+  }
+
+  npy_intp* array_shape = PyArray_DIMS(array);
+  for (int64_t i = 0; i < ndim; ++i) {
+    if (array_shape[i] != shape[i+start]) {
+      return false;
+    }
+  }
+  return true;
+}
+
+
+inline int _validate_ndarray(PyObject* obj, std::vector<int64_t> & shape, std::vector<int64_t> & ndarray_shape, int64_t depth) {
+  /*
+  This returns -1 if obj is a scalar array, 0 if obj is not array else if 1 given no exception is raised.
+  A ValueError will be raised if the depth of obj within the input iterable, as given to _extract_ndarrays,
+  is not the same as any other array within the input iterable OR if its shape does not match any other
+  array's shape in the iterable.
+  */
+
+  PyArrayObject* array;
+  if (PyArray_Check(obj)) {
+    array = (PyArrayObject*) obj;
+  } else {
+    return 0;
+  } 
+
+  if (PyArray_NDIM(array) == 0) {
+    return -1;
+  }
+
+  if (ndarray_shape.size() == 0) {
+    /*
+    If ndarray_shape is an empty vector, then ndarray_shape[0] is set to the depth of the array in the iterable and
+    ndarray_shape[1:] is set to the shape of array.
+    */
+    ndarray_shape.emplace_back(depth);
+    int64_t ndim = PyArray_NDIM(array);
+    npy_intp* array_shape = PyArray_DIMS(array);
+    ndarray_shape.insert(ndarray_shape.end(), array_shape, array_shape + ndim);
+  }
+  else if ((ndarray_shape[0] != depth) || !_shape_match_ndarray(ndarray_shape, array, 1)) {
```

Comment:
Use 

https://github.com/pytorch/pytorch/blob/5b648ef909fbf89c53f28ebc1b3bd2f4fde168c5/c10/util/Exception.h#L435

Reply:
Replaced it.